src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage_Threading.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.SQLite.v2
+{
+    internal partial class SQLitePersistentStorage
+    {
+        /// <summary>
+        /// Use a <see cref="ConcurrentExclusiveSchedulerPair"/> to simulate a reader-writer lock.
+        /// Read operations are performed on the <see cref="ConcurrentExclusiveSchedulerPair.ConcurrentScheduler"/>
+        /// and writes are performed on the <see cref="ConcurrentExclusiveSchedulerPair.ExclusiveScheduler"/>.
+        ///
+        /// We use this as a condition of using the in-memory shared-cache sqlite DB.  This DB
+        /// doesn't busy-wait when attempts are made to lock the tables in it, which can lead to
+        /// deadlocks.  Specifically, consider two threads doing the following:
+        ///
+        /// Thread A starts a transaction that starts as a reader, and later attempts to perform a
+        /// write. Thread B is a writer (either started that way, or started as a reader and
+        /// promoted to a writer first). B holds a RESERVED lock, waiting for readers to clear so it
+        /// can start writing. A holds a SHARED lock (it's a reader) and tries to acquire RESERVED
+        /// lock (so it can start writing).  The only way to make progress in this situation is for
+        /// one of the transactions to roll back. No amount of waiting will help, so when SQLite
+        /// detects this situation, it doesn't honor the busy timeout.
+        ///
+        /// To prevent this scenario, we control our access to the db explicitly with operations that
+        /// can concurrently read, and operations that exclusively write.
```

Comment:
So does this mean all the transaction handling elsewhere isn't really doing anything? How should we mentally compare the two?

Reply:
i'll update.  but it's subtle.  the transaction ensures both all-or-nothing *and* perf (the latter being the thing we care the most about).  However, transactions can 'fail' in sqlite with things like lock-taking-order.  I wrote up a full comment in this file to explain how that happens and i'm happy to clarify more.

However, the lock-reordering can only be an issue across writes  as reads don't ever promote their locks from read-locsk to write-locks.  so reads can all be concurrent safely.  but writes, unfortunately can never be concurrent (even with transactions).

The schedulers here are to prevent hte locking issue.  The transactions are to get the perf.